cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java

```diff
@@ -50,13 +59,27 @@
   @Override
   protected void append(LogMessage logMessage) {
     try {
+      String partitionKey = getPartitionKey(logMessage.getLoggingContext());
       byte [] bytes = loggingEventSerializer.toBytes(logMessage.getLoggingEvent(), logMessage.getLoggingContext());
-      producer.publish(logMessage.getLoggingContext().getLogPartition(), bytes);
+      producer.publish(partitionKey, bytes);
     } catch (Throwable t) {
       LOG.error("Got exception while serializing log event {}.", logMessage.getLoggingEvent(), t);
     }
   }
 
+  private String getPartitionKey(LoggingContext loggingContext) {
+    String namespaceId = loggingContext.getSystemTagsMap().get(NamespaceLoggingContext.TAG_NAMESPACE_ID).getValue();
+
+    if (cConf.get(Constants.Logging.LOG_PUBLISH_PARTITION_KEY).equals(PartitionType.PROGRAM.getType()) ||
```

Comment:
Also, it's worth to add a check in the `ConfigurationCheck` class to make validation using the `PartitionType.valueOf()` call.